core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -264,25 +264,31 @@ class KafkaApis(val requestChannel: RequestChannel,
       }.toMap
       sendResponseMaybeThrottle(request, requestThrottleMs => new OffsetCommitResponse(requestThrottleMs, results.asJava))
     } else {
-      val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) = offsetCommitRequest.offsetData.asScala.toMap.partition {
-        case (topicPartition, _) =>
-          val authorizedForDescribe = authorize(request.session, Describe, new Resource(Topic, topicPartition.topic))
-          val exists = metadataCache.contains(topicPartition.topic)
-          if (!authorizedForDescribe && exists)
-              debug(s"Offset commit request with correlation id ${header.correlationId} from client ${header.clientId} " +
-                s"on partition $topicPartition failing due to user not having DESCRIBE authorization, but returning UNKNOWN_TOPIC_OR_PARTITION")
-          authorizedForDescribe && exists
-      }
 
-      val (authorizedTopics, unauthorizedForReadTopics) = existingAndAuthorizedForDescribeTopics.partition {
-        case (topicPartition, _) => authorize(request.session, Read, new Resource(Topic, topicPartition.topic))
+      var unauthorizedForDescribeTopics = mutable.Map[TopicPartition, OffsetCommitRequest.PartitionData]()
+      var nonExistingAndAuthorizedForDescribeTopics = mutable.Map[TopicPartition, OffsetCommitRequest.PartitionData]()
+      var unauthorizedForReadTopics = mutable.Map[TopicPartition, OffsetCommitRequest.PartitionData]()
+      var authorizedTopics = mutable.Map[TopicPartition, OffsetCommitRequest.PartitionData]()
+
+      for ((topicPartition, partitionData) <- offsetCommitRequest.offsetData.asScala.toMap) {
+        if (authorize(request.session, Describe, new Resource(Topic, topicPartition.topic)))
+          if (metadataCache.contains(topicPartition.topic))
+            if (authorize(request.session, Read, new Resource(Topic, topicPartition.topic)))
+              authorizedTopics += (topicPartition -> partitionData)
+            else
+              unauthorizedForReadTopics += (topicPartition -> partitionData)
+          else
+            nonExistingAndAuthorizedForDescribeTopics += (topicPartition -> partitionData)
+        else
+          unauthorizedForDescribeTopics += (topicPartition -> partitionData)
```

Comment:
@hachikuji  I tried to implement using mutable collections  and single pass over the map.  If this is OK, or Can you suggest any nicer way to achive this?  Once you confirm, i will implement in other places.

Reply:
@ijuma Can you suggest a nicer approach for above code block?